src/expr/src/typecheck.rs

```diff
@@ -0,0 +1,774 @@
+// by the Apache License, Version 2.0.
+
+//! Thorough consistency checking and type synthesis for MIR expressions
+
+use itertools::Itertools;
+use mz_repr::{ColumnType, RelationType, Row, ScalarType};
+use std::collections::{BTreeMap, BTreeSet};
+use tracing::warn;
+
+use crate::{
+    func as scalar_func, AggregateExpr, ColumnOrder, Id, JoinImplementation, LocalId,
+    MirRelationExpr, MirScalarExpr, UnaryFunc,
+};
+
+/// The possible forms of inconsistency/errors discovered during typechecking.
+///
+/// Every variant has a `source` field identifying the MIR term that is home
+///  to the error (though not necessarily the root cause of the error).
+#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub enum TypeError<'a> {
+    Unbound {
+        source: &'a MirRelationExpr,
+        id: Id,
+        typ: RelationType,
+    },
+    NoSuchColumn {
+        source: &'a MirRelationExpr,
+        expr: &'a MirScalarExpr,
+        col: usize,
+    },
+    MismatchColumn {
+        source: &'a MirRelationExpr,
+        got: ColumnType,
+        expected: ColumnType,
+        message: String,
+    },
+    MismatchColumns {
+        source: &'a MirRelationExpr,
+        got: Vec<ColumnType>,
+        expected: Vec<ColumnType>,
+        message: String,
+    },
+    BadConstantRow {
+        source: &'a MirRelationExpr,
+        got: Row,
+        expected: Vec<ColumnType>,
+    },
+    BadProject {
+        source: &'a MirRelationExpr,
+        got: Vec<usize>,
+        input_type: Vec<ColumnType>,
+    },
+    BadTopKGroupKey {
+        source: &'a MirRelationExpr,
+        key: usize,
+        input_type: Vec<ColumnType>,
+    },
+    BadTopKOrdering {
+        source: &'a MirRelationExpr,
+        order: ColumnOrder,
+        input_type: Vec<ColumnType>,
+    },
+    BadLetRecBindings {
+        source: &'a MirRelationExpr,
+    },
+}
+
+type Ctx = BTreeMap<Id, Vec<ColumnType>>;
+
+/// Returns true when it is safe to treat a `got` row as an `expected` row
+///
+/// In particular, the core types must be equal, and if a column in `known` is nullable, that column should also be nullable in `got`
+/// Conversely, it is okay to treat a known non-nullable column as nullable: `got` may be nullable when `known` is not
+pub fn columns_match(got: &[ColumnType], known: &[ColumnType]) -> bool {
+    if got.len() != known.len() {
+        return false;
+    }
+
+    got.iter().zip_eq(known.iter()).all(|(got, known)| {
+        (!known.nullable || got.nullable) && got.scalar_type.base_eq(&known.scalar_type)
+    })
+}
+
+impl MirRelationExpr {
+    /// Returns the type of a relation expression or a type error.
+    ///
+    /// This function is careful to check validity, not just find out the type.
+    ///
+    /// It should be linear in the size of the AST.
+    ///
+    /// ??? should we also compute keys and return a `RelationType`?
+    pub fn typecheck(&self, ctx: &Ctx) -> Result<Vec<ColumnType>, TypeError> {
```

Comment:
Another thing that I noticed - at the moment the only use of the `typcheck` method is in the `Typecheck` transform. 

Do we expect this to change in the future? I think you might save up some levels of indirection (especially if you want to use a `RecursionGuard` to make the traversal properly handle large terms) if you just moved the `typecheck` method to the `Typecheck` transform.

Reply:
I'll be working on stack safety stuff tomorrow. Thanks for the pointers!

I put `typecheck` in `expr` at first because it's where `col_from_input_cols` was and it seemed like a natural enough place... but I see your point! I see two issues with moving the typechecker to `transform`, not sure how severe these are:

  1. Discoverability will be better in `expr`.
  2. Other users of the typechecker will have to depend on `transform`, not just `expr`.

Reply:
Thoughts on whether to move typechecking from `mz_expr` to `mz_transform`? I'm happy either way.